### DISCUSSION TITLE
Toggle dark mode stylex facebook

### DISCUSSION THREAD

**Discussion by felpereira** [facebook/stylex#511] [Q&A] [answered]
Sure, here's how you can create a button to toggle between dark mode and the default theme using the example below:

```TS
export const colors = stylex.defineVars({
  primaryText: {default: 'black', [DARK]: 'white'},
  secondaryText: {default: '#333', [DARK]: '#ccc'},
  accent: {default: 'blue', [DARK]: 'lightblue'},
  background: {default: 'white', [DARK]: 'black'},
  lineColor: {default: 'gray', [DARK]: 'lightgray'},
});
```

I wanted to make something like that button.

![image](https://github.com/facebook/stylex/assets/18452175/155dff6e-b158-4c5a-88dd-3a76c1400901)

Sorry, I've been trying all day already, I'm new to using StyleX and new to the frontend.

**Comment by nmn** [comment, marked as answer]
You'd need two themes as well.

```tsx
import {color} from './globalTheme.stylex'

export const light = stylex.createTheme(color, {
  primaryText: 'black',
  secondaryText: '#333',
  accent: 'blue',
  background: 'white',
  lineColor: 'gray',
});

export const dark = stylex.defineVars({
  primaryText: 'white',
  secondaryText: '#ccc',
  accent: 'lightblue',
  background: 'black',
  lineColor: 'lightgray',
});
```

Now, somewhere near the root of your app you can conditionally wrap everything with one of the two themes.


```tsx
import {light, dark} from './themes'

function Root({children}) {
  const [theme, setTheme] = useState('auto');

  return (
    <div {...stylex.props(
      theme === 'light' && light,
      theme === 'dark' && dark,
    )}>
      {children}
    </div>
  )
}
```

**Reply by TxHawks** [reply]
There isn't really any way to avoid a flash of light-mode with this approach doing SSR and persisting the user choice in local storage cause checking for the stored value would have to be done in `useEffect`, and setting the state would then only happen after hydration.

If we could somehow have dark and light mode based on a selector instead of a media query, and inline script in the `head` could check for a stored value and change a class or data attribute on the `html` element before any content is rendered.

It should be pretty easy to detect if the condition is an at-rule or a selector, and emit a ruleset with or without the enclosed `:root`.

**Reply by nmn** [reply]
The new `light-dark()` CSS function which is newly supported in all three browser engines can be used to make this all a LOT easier.

A single theme can be created with `light-dark(a, b)` used for all CSS variables. Then `color-scheme` can be set to `light`, `dark` or `light dark` to control the theme.

**Reply by TxHawks** [reply]
Yeah, light-dark is really nice, and would definitely solve this specific problem, but support is a little to new still... Only 65.84% global support according to caniuse.

**Reply by nmn** [reply]
If you're not using multiple layers of variables, LightningCSS can polyfill `light-dark` for you

**Comment by felpereira** [comment]
ENTENDI

I understood!

In my case, const Colors is the default color, and the themes are as I put them in the body

```ts
import * as stylex from "@stylexjs/stylex";

export const colors = stylex.defineVars({
  "gray-900": "#0C0C0D",
  "gray-800": "#131313",
  "gray-700": "#272727",
  "gray-400": "#6F6F6F",
  "gray-300": "#C8C8C8",
  "green-500": "#5FB9B0",
  "purple-300": "#B292FF",
  white: "#FFFFFF",
});

export const dark = stylex.createTheme(colors, {
  "gray-900": "#0C0C0D",
  "gray-800": "#131313",
  "gray-700": "#272727",
  "gray-400": "#6F6F6F",
  "gray-300": "#C8C8C8",
  "green-500": "#5FB9B0",
  "purple-300": "#B292FF",
  white: "#FFFFFF",
});

export const light = stylex.createTheme(colors, {
  "gray-900": "red",
  "gray-800": "red",
  "gray-700": "red",
  "gray-400": "red",
  "gray-300": "red",
  "green-500": "red",
  "purple-300": "red",
  white: "red",
});
```
APPLICATION

```ts
import type { Metadata } from "next";
import * as stylex from "@stylexjs/stylex";

import "./globals.css";
import { light, dark } from "@/modules/theme/tokens.stylex";

export const metadata: Metadata = {
  title: "Portfolio - Felipe Pereira",
};

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="pt-br">
      <head></head>
      <body {...stylex.props(dark)}>{children}</body>
    </html>
  );
}

```

**Reply by nmn** [reply]
@felpereira I suggest using camelCase for your themes as these are names used in JS source only.

i.e. `gray-900` -> `gray900`

Also, `createTheme` should ideally be in a separate file from `defineVars`.

**Reply by felpereira** [reply]
Thank you, yes I will do it this way, I will also test @predaytor  suggestion

**Comment by predaytor** [comment]
You can choose not to use themes at all if you only want browser's light/dark default. There is a new CSS property [light-dark()](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/light-dark) which allows you to avoid using `@media (prefers-color-scheme)`. Thanks to CSS preprocessors like [LightningCSS](https://lightningcss.dev/), you can use it today.

For convenience, I like to separate light and dark tokens. You can also refer to them to get more meaningful token names (as with regular CSS variables, which they are). Don't worry about duplicate tokens (in the case of aliases like `supportive`), gzip handles this gracefully.

`colors/light.stylex.ts`:
```tsx
import * as stylex from '@stylexjs/stylex';

const gray = {
	gray1: 'oklch(99.1% 0.001 277.7)',
	// ...
} as const;

// ...

const supportive = {
	supportive1: gray.gray1,
	// ...
};

export const light = stylex.defineVars({
	...gray,
	...supportive,
});
```
`colors/dark.stylex.ts`:
```tsx
import * as stylex from '@stylexjs/stylex';

const gray = {
	gray1: 'oklch(17.8% 0.004 277.7)',
	// ...
} as const;

// ...

const supportive = {
	supportive1: gray.gray1,
	// ...
};

export const dark = stylex.defineVars({
	...gray,
	...supportive,
});
```
Then merge them into a `light-dark()` function, selecting each one individually:

`colors/index.stylex.ts`:
```tsx
import * as stylex from '@stylexjs/stylex';

import { dark } from '~/styles/colors/dark.stylex';
import { light } from '~/styles/colors/light.stylex';

export const colors = stylex.defineVars({
	gray1: `light-dark(${light.gray1}, ${dark.gray1})`,
	// ...

	supportive1: `light-dark(${light.supportive1}, ${dark.supportive1})`,
} as const);
```
Switch the color-scheme programmatically or let the browser decide:
```tsx
import * as stylex from '@stylexjs/stylex';

export function Layout({ children }: { children: React.ReactNode }) {
	return (
		<html {...stylex.props(styles.html)}>
			<head>
				<meta charSet="utf-8" />
				<meta name="viewport" content="width=device-width, initial-scale=1" />
			</head>

			<body>{children}</body>
		</html>
	);
}

const styles = stylex.create({
	html: {
		// let the system decide which mode to use
		colorScheme: 'light dark',

		// or switch manually
		colorScheme: 'light',
		// colorScheme: 'dark',
	},
});
```

**Reply by nmn** [reply]
> Thanks to CSS preprocessors like [LightningCSS](https://lightningcss.dev/), you can use it today.

LightningCSS doesn't handle `light-dark()` correctly when it is used as a value of a CSS variable. I'm working to solve this.


---

You can do this with consts in the same file. If you're using `light-dark` the whole thing can be a single `defineVars()` call

colors/index.stylex.ts:
```tsx
import * as stylex from '@stylexjs/stylex';

const gray = {
	gray1: 'light-dark(oklch(99.1% 0.001 277.7), oklch(17.8% 0.004 277.7))',
	// ...
} as const;

const supportive = {
  supportive1: gray.gray1,
  // ...
};

export const colors = stylex.defineVars({
  gray1: dark.gray1,
  // ...
  supportive1: supportive.supportive1,
} as const);
```

You can define separate objects for light and dark in the same file if you want. You don't need multiple `defineVars` when using `light-dark()`. `color-scheme` should make it all work after the fact.

---

Side note: The `as const` within the `stylex.defineVars` should be handled by StyleX for you. Fixing this shortly.

**Reply by predaytor** [reply]
> You can define separate objects for light and dark in the same file if you want.

@nmn yeah, the idea was to be able to select individual colors (light or dark) to create a non-uniform light-dark() result, like so:
```tsx
export const colors = stylex.defineVars({
	...gray,
	canvas: `light-dark(#fff, ${dark.gray5})`,
});
```
I tested your approach with solid light-dark variables in a single file and it gave the exact result with an even smaller bundle size (*gzipped*)!
```tsx
export const colors = stylex.defineVars({
    ...gray,
    canvas: `light-dark(#fff, ${gray.gray5})`,
    // ^? (property) canvas: "light-dark(#fff, light-dark( oklch(91% 0.008 277.7), oklch(31.4% 0.009 277.7) ))"
});
```
I thought we couldn't nest `light-dark` within another `light-dark`, but it worked (both native and LightningCSS polyfill). Thx 🙌

**Reply by predaytor** [reply]
> Side note: The as const within the stylex.defineVars should be handled by StyleX for you. Fixing this shortly.

@nmn, is the feature included in the new StyleX 0.6.1 release? Very thanks.

**Reply by nmn** [reply]
It *should* work. I didn't test this specifically.

**Reply by predaytor** [reply]
@nmn, can we make this small fix to the `defineVars` interface to support the const assertion (add the `const` keyword)? Thanks.
https://github.com/facebook/stylex/blob/main/packages/stylex/src/StyleXTypes.d.ts#L197-L202
```ts
export type StyleX$DefineVars = <
  const DefaultTokens extends TTokens,
  ID extends symbol = symbol,
>(
  tokens: DefaultTokens,
) => VarGroup<FlattenTokens<DefaultTokens>, ID>;
```

**Reply by nmn** [reply]
`const` doesn't work as expected as the types enforce what can be passed in when creating a "theme" with `createTheme`. Using `const` keyword, would constrain the value types to the literals that they are initialised with.
